labelbox/data/metrics/confusion_matrix/calculation.py

```diff
@@ -0,0 +1,288 @@
+from ..group import (get_feature_pairs, get_identifying_key, has_no_annotations,
+                     has_no_matching_annotations)
+
+
+def confusion_matrix(ground_truths: List[Union[ObjectAnnotation,
+                                               ClassificationAnnotation]],
+                     predictions: List[Union[ObjectAnnotation,
+                                             ClassificationAnnotation]],
+                     include_subclasses: bool,
+                     iou: float) -> ConfusionMatrixMetricValue:
+    """
+    Computes the confusion matrix for an arbitrary set of ground truth and predicted annotations.
+    It first computes the confusion matrix for each metric and then sums across all classes
+
+    Args:
+        ground_truth : Label containing human annotations or annotations known to be correct
+        prediction: Label representing model predictions
+        include_subclasses (bool): Whether or not to include subclasses in the calculation.
+            If set to True, the iou between two overlapping objects of the same type is 0 if the subclasses are not the same.
+        iou: minimum overlap between objects for them to count as matching
+    Returns:
+        confusion matrix as a list: [TP,FP,TN,FN]
+        Returns None if there are no annotations in ground_truth or prediction annotations
+    """
+
+    annotation_pairs = get_feature_pairs(ground_truths, predictions)
+    conf_matrix = [
+        feature_confusion_matrix(annotation_pair[0], annotation_pair[1],
+                                 include_subclasses, iou)
+        for annotation_pair in annotation_pairs.values()
+    ]
+    matrices = [matrix for matrix in conf_matrix if matrix is not None]
+    return None if not len(matrices) else np.sum(matrices, axis=0).tolist()
+
+
+def feature_confusion_matrix(
+        ground_truths: List[Union[ObjectAnnotation, ClassificationAnnotation]],
+        predictions: List[Union[ObjectAnnotation, ClassificationAnnotation]],
+        include_subclasses: bool,
+        iou: float) -> Optional[ConfusionMatrixMetricValue]:
+    """
+    Computes confusion matrix for all features of the same class.
+
+    Args:
+        ground_truths: List of ground truth annotations belonging to the same class.
+        predictions: List of annotations  belonging to the same class.
+        include_subclasses (bool): Whether or not to include subclasses in the calculation.
+            If set to True, the iou between two overlapping objects of the same type is 0 if the subclasses are not the same.
+    Returns:
+        confusion matrix as a list: [TP,FP,TN,FN]
+        Returns None if there are no annotations in ground_truth or prediction annotations
+    """
+    if has_no_matching_annotations(ground_truths, predictions):
+        return [0, int(len(predictions) > 0), 0, int(len(ground_truths) > 0)]
+    elif has_no_annotations(ground_truths, predictions):
+        return None
+    elif isinstance(predictions[0].value, Mask):
```

Comment:
why do we only check the first value?

Reply:
Because we have already checked that they are all the same type previously. So we just need to quickly check again what is contained in the list.